<launch>
    <!-- Arguments -->
    <arg name="mode" default="sitl"/>
    <arg name="ns_prefix" default="uav_"/>
    <arg name="robot_model" default="iris"/>
    <arg name="pub_rate" default="100.0"/>
    <arg name="debug" default="false"/>
    <arg name="save_test_data" default="false"/>
    <arg name="save_experiment_data" default="false"/>
    <arg name="trajectory" default="true"/>
    <arg name="reach_tolerance" default="0.4"/>
    <arg name="path" default="optitrack2"/>
    <arg name="generator_mode" default="0"/>
    <arg name="pkg_name" default="ual_communication"/>
    <arg name="use_class" default="true"/>
    <arg name="multi" default="false"/>
    <arg name="cruising_speed" default="1.0"/>
    <arg name="look_ahead" default="0.7"/>
    <arg name="light" default="false"/>
    <arg name="sitl" default="true"/>

    <node pkg="rviz" type="rviz" name="rviz_node" args="-d $(find usp_testers)/config/tester.rviz" required="true"/>

    <include file="$(find ual_communication)/launch/sim_server_ual.launch" unless="$(arg light)">
        <arg name="multi" value="$(arg multi)"/>
        <arg name="robot_model" value="$(arg robot_model)"/>
    </include>
        <include file="$(find ual_communication)/launch/sim_server_ual_light.launch" if="$(arg light)">
        <arg name="multi" value="$(arg multi)"/>
        <!-- <arg name="robot_model" value="$(arg robot_model)"/> -->
    </include>

    <node pkg="db_manager" type="DBmanager" name="data_base_manager" output="screen" required="true"/>
    <node pkg="tactical_deconfliction" type="ConflictSolver" name="tactical_deconfliction" output="screen" required="true"/>
    <node pkg="monitoring" type="monitoring" name="monitoring"  required="true"/>
    <node pkg="usp_testers" type="solve_geofence" name="tester_1" output="screen" required="true"/>
    <node pkg="upat_follower" type="multi_uav_log_node" name="multi_uav_log" if="$(arg save_experiment_data)"/> 
    <!-- <node pkg="emergency_management" type="emergency_management.py" name="emergency_management" output="screen" required="true"/> -->

    <!-- <node pkg="tf2_ros" type="static_transform_publisher" name="tf_map_world" args="0 0 0 0 0 0 1 world map" /> -->

    <!-- UAV Path Manager nodes -->
    <group ns="upat_follower">
        <group ns="$(arg ns_prefix)0">
            <node pkg="$(arg pkg_name)" type="ualCommunication" name="ual_communication" output="screen" required="true">
                <param name="uav_id" value="0"/>
                <param name="ns_prefix" value="$(arg ns_prefix)"/>
                <param name="save_test_data" value="$(arg save_test_data)"/>
                <param name="trajectory" value="false"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
                <param name="path" value="$(arg path)"/>
                <param name="reach_tolerance" value="$(arg reach_tolerance)"/>
                <param name="use_class" value="true"/>
                <param name="generator_mode" value="$(arg generator_mode)"/>
                <param name="debug" value="$(arg debug)"/>
                <param name="cruising_speed" value="$(arg cruising_speed)"/>
                <param name="look_ahead" value="$(arg look_ahead)"/>
                <param name="light" value="$(arg light)"/>
                <param name="sitl" value="$(arg sitl)"/>
            </node>
            <node pkg="upat_follower" type="visualization_node" name="visualization" required="true" output="screen">
                <param name="uav_id" value="0"/>
                <param name="ns_prefix" value="$(arg ns_prefix)"/>
                <param name="robot_model" value="$(arg robot_model)"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
                <param name="save_experiment_data" value="$(arg save_experiment_data)"/>
                <param name="trajectory" value="$(arg trajectory)"/>
                <param name="generator_mode" value="$(arg generator_mode)"/>
            </node>
        </group>
        <group ns="$(arg ns_prefix)1">
            <node pkg="$(arg pkg_name)" type="ualCommunication" name="ual_communication" output="screen" required="true" if="$(arg multi)">
                <param name="uav_id" value="1"/>
                <param name="ns_prefix" value="$(arg ns_prefix)"/>
                <param name="save_test_data" value="$(arg save_test_data)"/>
                <param name="trajectory" value="$(arg trajectory)"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
                <param name="path" value="$(arg path)"/>
                <param name="reach_tolerance" value="$(arg reach_tolerance)"/>
                <param name="use_class" value="true"/>
                <param name="generator_mode" value="$(arg generator_mode)"/>
                <param name="debug" value="$(arg debug)"/>
                <param name="cruising_speed" value="$(arg cruising_speed)"/>
                <param name="look_ahead" value="$(arg look_ahead)"/>
                <param name="light" value="$(arg light)"/>
                <param name="sitl" value="$(arg sitl)"/>
            </node>
            <node pkg="upat_follower" type="visualization_node" name="visualization" required="true" output="screen">
                <param name="uav_id" value="1"/>
                <param name="ns_prefix" value="$(arg ns_prefix)"/>
                <param name="robot_model" value="$(arg robot_model)"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
                <param name="save_experiment_data" value="$(arg save_experiment_data)"/>
                <param name="trajectory" value="$(arg trajectory)"/>
                <param name="generator_mode" value="$(arg generator_mode)"/>
            </node>
        </group>
        <group ns="$(arg ns_prefix)2">
            <node pkg="$(arg pkg_name)" type="ualCommunication" name="ual_communication" output="screen" required="true" if="$(arg multi)">
                <param name="uav_id" value="2"/>
                <param name="ns_prefix" value="$(arg ns_prefix)"/>
                <param name="save_test_data" value="$(arg save_test_data)"/>
                <param name="trajectory" value="$(arg trajectory)"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
                <param name="path" value="$(arg path)"/>
                <param name="reach_tolerance" value="$(arg reach_tolerance)"/>
                <param name="use_class" value="true"/>
                <param name="generator_mode" value="$(arg generator_mode)"/>
                <param name="debug" value="$(arg debug)"/>
                <param name="cruising_speed" value="$(arg cruising_speed)"/>
                <param name="look_ahead" value="$(arg look_ahead)"/>
                <param name="light" value="$(arg light)"/>
                <param name="sitl" value="$(arg sitl)"/>
            </node>
            <node pkg="upat_follower" type="visualization_node" name="visualization" required="true" output="screen">
                <param name="uav_id" value="2"/>
                <param name="ns_prefix" value="$(arg ns_prefix)"/>
                <param name="robot_model" value="$(arg robot_model)"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
                <param name="save_experiment_data" value="$(arg save_experiment_data)"/>
                <param name="trajectory" value="$(arg trajectory)"/>
                <param name="generator_mode" value="$(arg generator_mode)"/>
            </node>
        </group>
    </group>

</launch>
